### ISSUE TITLE
RedisValue equality and GetHashCode

### ISSUE THREAD

**Issue by cmoncure** [StackExchange/StackExchange.Redis#733]
One thing I would like to do is to compare a Dictionary<RedisValue, object> with the HashEntry[] returned by e.g. IDatabase.HashGetAll().  I run into a problem if I key my objects with an integer backed RedisValue.    Things look okay until you persist to the Redis hash, but when they are retrieved, the RedisValue hash field keys come back as strings, ~~and now inMemoryValue == retrievedValue no longer holds.~~

~~I have an idea to alter RedisValue so that equality checks if RedisValue this with a valueInt64 has an equivalent representation in RedisValue Other with a valueBlob.  Redis itself demotes stored integer values to strings in both the hash field and hash value, so it ought to be semantically correct. Would this break anything?~~

(update:)

Actually, now that I dug into the code and tested, it looks like my issue is a discrepancy between the equality overrides and the GetHashCode override.

```
        public void TestRedisValueSetMembership() {
            RedisValue A_int = 1;
            RedisValue A_string = "1";
            Assert.Equal(A_int, A_string);

            var set = new HashSet<RedisValue> { A_int };
            Assert.True(set.Contains(A_string));
        } 
```

This is failing because in GetHashCode() we have this:

`            if (valueBlob == IntegerSentinel) return valueInt64.GetHashCode(); `

Whereas in operator == we have this:

```
            else if (y.valueBlob == IntegerSentinel)
            {
                return Equals((byte[])x, (byte[])y);
            }
```

and 
```
            if (valueBlob == IntegerSentinel)
            {
                return Encoding.UTF8.GetBytes(Format.ToString(value.valueInt64));
            }
```

So for equality we are testing the converted string value, but for GetHashCode we are just returning the hash code of the backing integer.  It makes sense to me for value types to have identity when there is equality, even more so in this case because Redis itself treats 1 and "1" as equal and identity, but will this break something for some users?

**Comment by cmoncure** [comment]
Actually, now that I dug into the code and tested, it looks like my issue is a discrepancy between the equality overrides and the GetHashCode override.

```
        public void TestRedisValueSetMembership() {
            RedisValue A_int = 1;
            RedisValue A_string = "1";
            Assert.Equal(A_int, A_string);

            var set = new HashSet<RedisValue> { A_int };
            Assert.True(set.Contains(A_string));
        } 
```

This is failing because in GetHashCode() we have this:

`            if (valueBlob == IntegerSentinel) return valueInt64.GetHashCode(); `

Whereas in operator == we have this:

```
            else if (y.valueBlob == IntegerSentinel)
            {
                return Equals((byte[])x, (byte[])y);
            }
```

and 
```
            if (valueBlob == IntegerSentinel)
            {
                return Encoding.UTF8.GetBytes(Format.ToString(value.valueInt64));
            }
```

**Comment by mgravell** [comment]
that discrepancy seems odd; will try to look - agree that sounds reasonable
(before looking, obviously)

**Comment by cmoncure** [comment]
Thanks for your attention to this

**Comment by mgravell** [comment]
This is fixed in 2.0: closing; please see https://github.com/StackExchange/StackExchange.Redis/issues/871
